core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -753,8 +763,15 @@ class KafkaServer(
             }
           }
           if (!shutdownSucceeded) {
-            Thread.sleep(config.controlledShutdownRetryBackoffMs)
-            warn("Retrying controlled shutdown after the previous attempt failed...")
+            var attempts = retries - remainingRetries
+            if (attempts < 1) {
```

Comment:
I'd prefer to just remove this if check if it would never happen given the current implementation. If we want to left some notes for the following developers, we may as well just add some comments saying "make sure this attempts value is always > 1 by checking XXX".

Reply:
I prefer to be safer here to prevent potential silently more frequent retries (negative value will give a smaller retry interval). It is not a heavy operation, and the retry rate is low (1 request every > 10 seconds based on the current config), so this check will not bring much burden to the controller.